### PR TITLE
EGM-certified electron IDs for 2016 data: cut-based and MVA

### PR DESCRIPTION
**Pull request by ikrav**
This PR contains EGM certified cut-based and MVA electron IDs tuned on 80X samples. They are intended to be used on 2016 data. The electron MVA contains the HZZ MVA used for ICHEP16 and the presently recommended general purpose MVA tuned for better performance given features of a more complete 2016 dataset.

This PR is a new version of the recent PR #16839 (about to be closed) that is rebased to CMSSW_9_0_X_2016-12-07-2300. It also contains a few minor changes addressing  comments from reviewers.

### CODE REVIEW

File: RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+from PhysicsTools.SelectorUtils.centralIDRegistry import central_id_registry
+
+# Common functions and classes for ID definition are imported here:
+from RecoEgamma.ElectronIdentification.Identification.cutBasedElectronID_tools import *
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py]: Still one `import *` in the new code -- please fix like you did the other one. Thanks.

**Comment by slava77** [comment]
On 12/12/16 12:33 PM, Carl Vuosalo wrote:
> *@cvuosalo* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In
> RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py
> <https://github.com/cms-sw/cmssw/pull/16979#pullrequestreview-12543328>:
> 
>> @@ -0,0 +1,99 @@
> +from PhysicsTools.SelectorUtils.centralIDRegistry import central_id_registry
> +
> +# Common functions and classes for ID definition are imported here:
> +from RecoEgamma.ElectronIdentification.Identification.cutBasedElectronID_tools import *
> 
> Still one |import *| in the new code -- please fix like you did the
> other one. Thanks.


This import *
cleanup is needed primarily for imports of files which define actual
modules.

It is not needed and is rather tedious to maintain for includes of
functions tools and ingredient PSets.

**Comment by ikrav** [comment]
Well, anyways I fixed it already, the list of explicit imports this time is 
not too long. The change is committed.

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16HZZ.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  void setConsumes(edm::ConsumesCollector&&) const override final;
+  // Call this function once per event to retrieve all needed
+  // event content pices
+  
+ private:
+
+  // MVA name. This is a unique name for this MVA implementation.
+  // It will be used as part of ValueMap names.
+  // For simplicity, keep it set to the class name.
+  const std::string _name = "ElectronMVAEstimatorRun2Spring16HZZ";
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16HZZ.h]: More prohibited names with leading underbars, lines 132-152.

File: RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleCalPFClusterIsoCut.cc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  double value(const reco::CandidatePtr& cand) const override final;
+
+  CandidateType candidateType() const override final { 
+    return ELECTRON; 
+  }
+
+private:
+  // Cut values
+  const float _isoCutEBLowPt,_isoCutEBHighPt,_isoCutEELowPt,_isoCutEEHighPt;
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleCalPFClusterIsoCut.cc]: Again, leading underbars lines 26-31.

File: RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ private:
+
+  // MVA name. This is a unique name for this MVA implementation.
+  // It will be used as part of ValueMap names.
+  // For simplicity, keep it set to the class name.
+  const std::string _name = "ElectronMVAEstimatorRun2Spring16GeneralPurpose";
+  // MVA tag. This is an additional string variable to distinguish
+  // instances of the estimator of this class configured with different
+  // weight files.
+  const std::string _tag;
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h]: There are still prohibited variables names that start with "_" here and on lines 136-149.

**Comment by ikrav** [comment]
Carl, so I made a number of changes in the new files, including the ones you
pointed out plus a few more, I think. The changes are pushed to GitHub.

However, let me mention: the present set of files new to CMSSW use files from 
older IDs as a starter. There is a ton of those old files in RecoEgamma. Most written by
Lindsey, actually. They are full of leading underscores. I have added some IDs in the past 
in the same style, and  nobody complained until now… I changed only the new files, but 
have not touched any of the old files.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ikrav** [comment]
The std::move was there in all previous electron and photon MVA IDs, are we required 
to do anything regarding it this time?

The error of the initialization: I am asking about it the people who wrote this part.
This whole constructor is new in comparison to previous ID versions, so I am not
sure what they intended here. My local build was successful before I submitted the PR,
but it wasn’t a clang build.

**Comment by ikrav** [comment]
The error below is fixed, the change is committed. The cause was my typo when
changing very many variable names to comply CMS coding rules in the previous commit.

File: RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleNormalizedGsfChi2Cut.cc [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  result_type operator()(const reco::GsfElectronPtr&) const override final;
+
+  double value(const reco::CandidatePtr& cand) const override final;
+
+  CandidateType candidateType() const override final { 
+    return ELECTRON; 
+  }
+
+private:
+  const double _normalizedGsfChi2CutValueEB,_normalizedGsfChi2CutValueEE,_barrelCutOff;
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleNormalizedGsfChi2Cut.cc]: Three more with prohibited leading underbars. Please fix them.

**Comment by ikrav** [comment]
This is fixed and committed.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #16979 8424971c2c661ab4651bfd98eb833cb1f5f97129

For EGM, adding new certified electron IDs.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_9_0_X_2016-12-15-2300 show no significant differences. An extended test for the previous version of this PR (https://github.com/cms-sw/cmssw/pull/16839#issuecomment-265842966), which differed only slightly from the current one, showed no problems, with no differences and only a very marginal increase in CPU time.

**Comment by ikrav** [comment]
Does this means I may proceed with submitting pull requests to 80X and 81X
with identical content? - thanks

**Comment by slava77** [comment]
this PR is not merged yet.
You can make backports, but there is a small chance you'd need to edit here and in the backports later.

**Comment by ikrav** [comment]
Ok, I take it that I shouldn’t submit new PRs until this PR is merged.
(I sent the email because the previous message said it is fully signed off on).
Higgs people asked me not to do backport to 80X, so instead I will cherry-pick
the commits you asked me to introduce in this PR, and add them to my original
80X branch.

**Comment by davidlange6** [comment]
+1

**Pull request merged**